File: c-sharp/CodifyTestHarness/1. Iterations/BinaryGap.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CodalityTestLibrary
{
    /// <summary>
    /// Write a function that, given a positive integer N, returns the length 
    /// of its longest binary gap.The function should return 0 if N doesn't 
    /// contain a binary gap.
    /// </summary>
    public class BinaryGap
    {
        public int solution(int number)
        {
            // Set the default return and get a binary representation of the number arguement
            int minimalReturnValue = 0;

            // If we get a zero then we can't do anything
            if(number == 0){
                return minimalReturnValue;
            }

            // Convert to binary
            string binaryRepresentation = Convert.ToString(number, 2);

            // Remove trailing zeros
            binaryRepresentation = binaryRepresentation.Remove(binaryRepresentation.LastIndexOf('1'));

            // Split the binary string down into a list of strings (1's and 0's)
            IEnumerable<string> binaryGaps = binaryRepresentation.Aggregate(" ", (x, y) => x + (x.Last() == y ? "" : " ") + y)
                                                          .Trim()
                                                          .Split(' ');

            // Check if we have split more than once, if not bail out and return 0
            if(binaryGaps.Count() <= 1)
            {
                return minimalReturnValue;
            }

            // iterate over the list and get the longest string
            int derivedReturnValue = binaryGaps.Aggregate((x, y) => x.Length > y.Length ? x : y).Length;

            // Return a positive value
            return derivedReturnValue; 

        }
    }
}

File: Python/1. Iterations/BinaryGap.py
import itertools

# Write a function that, given a positive integer N, returns the length 
# of its longest binary gap.The function should return 0 if N doesn't 
# contain a binary gap.
class BinaryGap(object):

    def solution(self, Number):

        # set retun value
        min_return_value = 0

        # gaurd against zero being passed
        if Number < 1:
            return min_return_value

        # make the numeber binary 11100101
        binaryRepresentation = str(bin(Number))[2:]
        
        # remove any trailing zeros
        binaryRepresentation = str(binaryRepresentation[:binaryRepresentation.rfind('1') + 1])

        # split the string by 1's and 0's
        grouping = [''.join(g) for k, g in itertools.groupby(binaryRepresentation)]

        # return the number of groups 
        if len(grouping) <= 1:
            return min_return_value
        
        # return the largest grouping as this (apparently) will equal the gaps
        return len(max(grouping, key=len))